lib/semantics/resolve-labels.cc

```diff
@@ -0,0 +1,969 @@
+    const B &labels, const C &scopes, ErrorHandler &eh) {
+  IndexList loopBodies;
+  for (const auto stmt : dos) {
+    const auto &label{std::get<parser::Label>(stmt)};
+    const auto &scope{std::get<ScopeProxy>(stmt)};
+    const auto &position{std::get<parser::CharBlock>(stmt)};
+    auto doTarget{GetLabel(labels, label)};
+    if (HasNoScope(std::get<ScopeProxy>(doTarget))) {
+      // C1133: this is an error: label not found
+      Report(eh, position,
+          parser::MessageFormattedText{
+              "label '%lu' cannot be found"_err_en_US, label});
+    } else if (std::get<parser::CharBlock>(doTarget).begin() <
+        position.begin()) {
+      // R1119: this is an error: label does not follow DO
+      Report(eh, position,
+          parser::MessageFormattedText{
+              "label '%lu' doesn't lexically follow DO stmt"_err_en_US, label});
+    } else if (!InInclusiveScope(
+                   scopes, scope, std::get<ScopeProxy>(doTarget))) {
+      // C1133: this is an error: label is not in scope
+      if (isStrictF18) {
+        Report(eh, position,
+            parser::MessageFormattedText{
+                "label '%lu' is not in scope"_err_en_US, label});
+      } else {
+        Report(eh, position,
+            parser::MessageFormattedText{
+                "label '%lu' is not in scope"_en_US, label});
+      }
+    } else if ((std::get<TargetStmtType>(doTarget) &
+                   TargetStmtType{TargetStatementEnum::Do})
+                   .none()) {
+      Report(eh, std::get<parser::CharBlock>(doTarget),
+          parser::MessageFormattedText{
+              "'%lu' invalid DO terminal statement"_err_en_US, label});
+    } else {
+      // save the loop body marks
+      loopBodies.push_back(
+          {SkipLabel(position), std::get<parser::CharBlock>(doTarget)});
+    }
+  }
+
+  // check that nothing jumps into the block
+  CheckBranchesIntoDoBody(branches, labels, scopes, loopBodies, eh);
+  // check that do loops properly nest
+  CheckDoNesting(loopBodies, eh);
+}
+
+/// \brief General constraint, control transfers within inclusive scope
+/// See, for example, section 6.2.5.
+template<typename A, typename B, typename C>
+void CheckScopeConstraints(
+    const A &stmts, const B &labels, const C &scopes, ErrorHandler &eh) {
+  for (const auto stmt : stmts) {
+    const auto &label{std::get<parser::Label>(stmt)};
+    const auto &scope{std::get<ScopeProxy>(stmt)};
+    const auto &position{std::get<parser::CharBlock>(stmt)};
+    auto target{GetLabel(labels, label)};
+    if (HasNoScope(std::get<ScopeProxy>(target))) {
+      // this is an error: label not found
+      Report(eh, position,
+          parser::MessageFormattedText{
+              "label '%lu' was not found"_err_en_US, label});
+    } else if (!InInclusiveScope(scopes, scope, std::get<ScopeProxy>(target))) {
+      // this is an error: label not in scope
+      if (isStrictF18) {
```

Comment:
See previous comments about errors vs warnings and the use of `%lu`.

(Pausing review here, will look at the rest of this code later.)